In [1]:
# Initial imports
import pandas as pd
from path import Path
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from sqlalchemy import create_engine
# from config import db_password

In [2]:
# Loading data
file_path = Path("WW_clean_data_labelEncoded_Winner.csv")
ww_df = pd.read_csv(file_path)
ww_df.head()

R_fighter       B_fighter       date  Winner  \
0            Song Kenan     Max Griffin  3/20/2021       0   
1          Leon Edwards  Belal Muhammad  3/13/2021       1   
2  Matthew Semelsberger      Jason Witt  3/13/2021       2   
3            Sean Brady   Jake Matthews   3/6/2021       2   
4        Belal Muhammad     Dhiego Lima  2/13/2021       2   

   B_total_time_fought(seconds)  B_total_rounds_fought  B_total_title_bouts  \
0                      0.584908                     28                    0   
1                      0.634631                     34                    0   
2                      0.158087                      3                    0   
3                      0.623774                     34                    0   
4                      0.548099                     18                    2   

   B_current_win_streak  B_current_lose_streak  B_longest_win_streak  ...  \
0                     0                      1                     1  ...   
1                     0                      1                     4  ...   
2                     0                      1                     1  ...   
3                     2                      0                     3  ...   
4                     0                      1                     3  ...   

   R_Country_Iceland  R_Country_New Zealand  R_Country_Nigeria  \
0                0.0                    0.0                0.0   
1                0.0                    0.0                0.0   
2                0.0                    0.0                0.0   
3                0.0                    0.0                0.0   
4                0.0                    0.0                0.0   

   R_Country_Palestine  R_Country_Russia  R_Country_USA  R_Country_brazil  \
0                  0.0               0.0            0.0               0.0   
1                  0.0               0.0            0.0               0.0   
2                  0.0               0.0            1.0               0.0   
3                  0.0               0.0            1.0               0.0   
4                  1.0               0.0            0.0               0.0   

   R_Country_canada  R_Country_england   R_Country_ireland  
0               0.0                 0.0                0.0  
1               0.0                 1.0                0.0  
2               0.0                 0.0                0.0  
3               0.0                 0.0                0.0  
4               0.0                 0.0                0.0  

[5 rows x 104 columns]

In [3]:
ww_df=ww_df.drop(['date'], axis=1)
ww_df.head()

R_fighter       B_fighter  Winner  B_total_time_fought(seconds)  \
0            Song Kenan     Max Griffin       0                      0.584908   
1          Leon Edwards  Belal Muhammad       1                      0.634631   
2  Matthew Semelsberger      Jason Witt       2                      0.158087   
3            Sean Brady   Jake Matthews       2                      0.623774   
4        Belal Muhammad     Dhiego Lima       2                      0.548099   

   B_total_rounds_fought  B_total_title_bouts  B_current_win_streak  \
0                     28                    0                     0   
1                     34                    0                     0   
2                      3                    0                     0   
3                     34                    0                     2   
4                     18                    2                     0   

   B_current_lose_streak  B_longest_win_streak  B_wins  ...  \
0                      1                     1       4  ...   
1                      1                     4       9  ...   
2                      1                     1       1  ...   
3                      0                     3      10  ...   
4                      1                     3       4  ...   

   R_Country_Iceland  R_Country_New Zealand  R_Country_Nigeria  \
0                0.0                    0.0                0.0   
1                0.0                    0.0                0.0   
2                0.0                    0.0                0.0   
3                0.0                    0.0                0.0   
4                0.0                    0.0                0.0   

   R_Country_Palestine  R_Country_Russia  R_Country_USA  R_Country_brazil  \
0                  0.0               0.0            0.0               0.0   
1                  0.0               0.0            0.0               0.0   
2                  0.0               0.0            1.0               0.0   
3                  0.0               0.0            1.0               0.0   
4                  1.0               0.0            0.0               0.0   

   R_Country_canada  R_Country_england   R_Country_ireland  
0               0.0                 0.0                0.0  
1               0.0                 1.0                0.0  
2               0.0                 0.0                0.0  
3               0.0                 0.0                0.0  
4               0.0                 0.0                0.0  

[5 rows x 103 columns]

In [4]:
# Define features and targets
X=ww_df.drop(['Winner', 'R_fighter', 'B_fighter'], axis=1)
y=ww_df.Winner.values.reshape(-1, 1)

In [5]:
#Splitting into Train(80%) and Test(20%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78, train_size=0.80)

In [6]:
# Creating a StandardScaler instance.
scaler = StandardScaler()
# Fitting the Standard Scaler with the training data.
X_scaler = scaler.fit(X_train)

# Scaling the data.
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [7]:
# Create a random forest classifier.
#n_estimator should be btwn 64 and 128 random forests
rf_model = RandomForestClassifier(n_estimators=128, random_state=78) 

In [8]:
# Fitting the model
rf_model = rf_model.fit(X_train_scaled, y_train)

C:\Users\kimbe\anaconda3\envs\mlenv\lib\site-packages\ipykernel_launcher.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  


In [9]:
# Making predictions using the testing data.
predictions = rf_model.predict(X_test_scaled)

In [10]:
# Calculating the confusion matrix.
cm = confusion_matrix(y_test, predictions)

# Create a DataFrame from the confusion matrix.
cm_df = pd.DataFrame(
    cm, index=["Actual 1", "Actual 2", "Actual 3"], columns=["Predicted 1", "Predicted 2", "Predicted 3"])

cm_df

Predicted 1  Predicted 2  Predicted 3
Actual 1            0            0            7
Actual 2            0            0            1
Actual 3            2            0           13

In [11]:
acc_score = accuracy_score(y_test, predictions)

print(f"Accuracy Score : {acc_score}")

Accuracy Score : 0.5652173913043478


In [12]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.00      0.00      0.00         7
           1       0.00      0.00      0.00         1
           2       0.62      0.87      0.72        15

    accuracy                           0.57        23
   macro avg       0.21      0.29      0.24        23
weighted avg       0.40      0.57      0.47        23



C:\Users\kimbe\anaconda3\envs\mlenv\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\kimbe\anaconda3\envs\mlenv\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\kimbe\anaconda3\envs\mlenv\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Accuracy Score : 0.5652173913043478
    
         precision    recall  f1-score   support

           0       0.00      0.00      0.00         7
           1       0.00      0.00      0.00         1
           2       0.62      0.87      0.72        15

    accuracy                           0.57        23
   macro avg       0.21      0.29      0.24        23
weighted avg       0.40      0.57      0.47        23

In [13]:
# Calculate feature importance in the Random Forest model.
importances = rf_model.feature_importances_
importances

array([0.04867049, 0.04202444, 0.0038042 , 0.01893708, 0.00843236,
       0.02224329, 0.02961708, 0.02974736, 0.        , 0.00109517,
       0.00589162, 0.02616968, 0.02708151, 0.01846929, 0.00195316,
       0.02790518, 0.03905924, 0.00022877, 0.01453639, 0.00723245,
       0.00022192, 0.00222106, 0.00046512, 0.00579547, 0.00115687,
       0.01173425, 0.00424349, 0.        , 0.00197716, 0.0011469 ,
       0.0039361 , 0.00056468, 0.03963786, 0.03555046, 0.01514303,
       0.01884613, 0.00724306, 0.03169521, 0.04536284, 0.02743781,
       0.        , 0.01075258, 0.01173775, 0.03986137, 0.02890074,
       0.02194462, 0.00702568, 0.01855029, 0.02197206, 0.00585984,
       0.02991984, 0.05087074, 0.00584627, 0.00697842, 0.01158523,
       0.00115479, 0.00720514, 0.00710212, 0.00240754, 0.        ,
       0.00062569, 0.00203326, 0.00163644, 0.00368979, 0.        ,
       0.00377384, 0.00345602, 0.00032167, 0.00350689, 0.0018845 ,
       0.0013849 , 0.        , 0.0004953 , 0.00498251, 0.00247

In [14]:
sorted(zip(rf_model.feature_importances_, X.columns), reverse=True)

[(0.05087074439485979, 'R_age'),
 (0.04867049006065448, 'B_total_time_fought(seconds)'),
 (0.04536284316216314, 'R_wins'),
 (0.04202444368538263, 'B_total_rounds_fought'),
 (0.03986136940746471, 'R_win_by_Decision_Unanimous'),
 (0.03963786459718838, 'R_total_time_fought(seconds)'),
 (0.0390592364112354, 'B_Reach_cms'),
 (0.035550462567240086, 'R_total_rounds_fought'),
 (0.031695212965861835, 'R_longest_win_streak'),
 (0.029919836002172448, 'B_age'),
 (0.029747363922810886, 'B_losses'),
 (0.02961708189622735, 'B_wins'),
 (0.02890073579983661, 'R_win_by_KO/TKO'),
 (0.027905176786684676, 'B_Height_cms'),
 (0.027437813143813786, 'R_losses'),
 (0.0270815050555423, 'B_win_by_KO/TKO'),
 (0.026169676906918793, 'B_win_by_Decision_Unanimous'),
 (0.022243293288136224, 'B_longest_win_streak'),
 (0.02197206496483825, 'R_Reach_cms'),
 (0.021944619537016832, 'R_win_by_Submission'),
 (0.018937082173223353, 'B_current_win_streak'),
 (0.018846127186819394, 'R_current_win_streak'),
 (0.018550292127186962

#Top three features
(0.05087074439485979, 'R_age'),
 (0.04867049006065448, 'B_total_time_fought(seconds)'),
 (0.04536284316216314, 'R_wins'),